File: CodingInterviewsPythonSolutions/155. Min Stack.py
class MinStack:

    def __init__(self):
        self.data = []
        self.mnStack = []

    def push(self, val: int) -> None:
        self.data.append(val)
        if not self.mnStack or val <= self.mnStack[-1]:
            self.mnStack.append(val)

    def pop(self) -> None:
        if self.mnStack[-1] == self.data[-1]:
            self.mnStack.pop()
        self.data.pop()

    def top(self) -> int:
        return self.data[-1]

    def getMin(self) -> int:
        return self.mnStack[-1]

File: CodingInterviewsJavaSolutions/155. Min Stack.java
class MinStack {
    private Stack<Integer> data;
    private Stack<Integer> mnStack;

    public MinStack() {
        data = new Stack<>();
        mnStack = new Stack<>();
    }

    public void push(int val) {
        data.push(val);
        if (mnStack.isEmpty() || val <= mnStack.peek()) {
            mnStack.push(val);
        }
    }

    public void pop() {
        if (mnStack.peek().equals(data.peek())) {
            mnStack.pop();
        }
        data.pop();
    }

    public int top() {
        return data.peek();
    }

    public int getMin() {
        return mnStack.peek();
    }
}
